pkg/piecestore/rpc/server/server.go

```diff
@@ -4,132 +4,126 @@
+	node := proto.Node{
+		Id: id,
+		Address: &proto.NodeAddress{
+			Transport: proto.NodeTransport_TCP,
+			Address:   kadAddress,
+		},
 	}
 
-	defer func() {
-		if err := storeFile.Close(); err != nil {
-			log.Printf("failed to close store file: %s\n", err)
-		}
-	}()
-
-	reader := NewStreamReader(stream)
-	total, err := io.Copy(storeFile, reader)
+	kad, err := kademlia.NewKademlia(kademlia.StringToNodeID(id), []proto.Node{node}, ip, kadListenPort)
 	if err != nil {
-		if err := cleanup(s, piece.Id); err != nil {
-			log.Printf("Failed on cleanup in Store: %s", err.Error())
-		}
+		return nil, errs.New("Failed to instantiate new Kademlia: %s", err.Error())
+	}
 
-		return err
+	if err := kad.ListenAndServe(); err != nil {
+		return nil, errs.New("Failed to ListenAndServe on new Kademlia: %s", err.Error())
 	}
 
-	log.Println("Successfully stored data.")
+	if err := kad.Bootstrap(ctx); err != nil {
+		return nil, errs.New("Failed to Bootstrap on new Kademlia: %s", err.Error())
+	}
 
-	return stream.SendAndClose(&pb.PieceStoreSummary{Message: OK, TotalReceived: int64(total)})
+	return kad, nil
 }
 
-// Retrieve -- Retrieve data from piecestore and send to client
-func (s *Server) Retrieve(pieceMeta *pb.PieceRetrieval, stream pb.PieceStoreRoutes_RetrieveServer) error {
-	log.Println("Retrieving data...")
+// Start the piececstore node
+func (s *Server) Start() error {
+	ctx := context.Background()
 
-	path, err := pstore.PathByID(pieceMeta.Id, s.PieceStoreDir)
+	_, err := connectToKad(ctx, s.config.NodeID, s.config.PsHost, s.config.KadListenPort, fmt.Sprintf("%s:%s", s.config.KadHost, s.config.KadPort))
 	if err != nil {
 		return err
 	}
 
-	fileInfo, err := os.Stat(path)
+	// create a listener on TCP port
+	lis, err := net.Listen("tcp", fmt.Sprintf(":%s", s.config.PsPort))
 	if err != nil {
 		return err
 	}
 
-	// Read the size specified
-	totalToRead := pieceMeta.Size
-	// Read the entire file if specified -1
-	if pieceMeta.Size <= -1 {
-		totalToRead = fileInfo.Size()
-	}
+	defer utils.Close(lis)
 
-	storeFile, err := pstore.RetrieveReader(stream.Context(), pieceMeta.Id, pieceMeta.Offset, totalToRead, s.PieceStoreDir)
-	if err != nil {
-		return err
-	}
+	// create a gRPC server object
+	grpcServer := grpc.NewServer()
 
-	defer func() {
-		if err := storeFile.Close(); err != nil {
-			log.Printf("failed to close store file: %s\n", err)
-		}
+	// attach the api service to the server
+	pb.RegisterPieceStoreRoutesServer(grpcServer, s)
+
+	// routinely check DB and delete expired entries
+	go func() {
+		err := s.DB.DBCleanup(s.DataDir)
+		zap.S().Fatalf("Error in DBCleanup: %v\n", err)
 	}()
 
-	writer := &StreamWriter{stream: stream}
-	_, err = io.Copy(writer, storeFile)
-	if err != nil {
-		return err
+	fmt.Printf("Node %s started\n", s.config.NodeID)
+
+	// start the server
+	if err := grpcServer.Serve(lis); err != nil {
+		zap.S().Fatalf("failed to serve: %s\n", err)
```

Comment:
`zap.S().Fatalf` will exit the program. We should just log the error and return it to the caller.